File: backend/app/services/websocket_service.py
import socketio
import logging
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)

# Create Socket.IO server with async support
sio = socketio.AsyncServer(
    async_mode='asgi',
    cors_allowed_origins='*',  # Allow all origins for development
    logger=True,
    engineio_logger=True
)

# Create ASGI app
socket_app = socketio.ASGIApp(sio)


class WebSocketService:
    """
    WebSocket service for real-time updates.
    
    Features:
    - Real-time notification delivery
    - Live validation count updates
    - Live consensus percentage updates
    - Badge achievement notifications
    - Trust score updates
    """
    
    def __init__(self):
        self.sio = sio
        self.connected_users = {}  # Maps user_id to session_id
    
    async def emit_to_user(self, user_id: str, event: str, data: Dict[str, Any]):
        """
        Emit an event to a specific user.
        
        Args:
            user_id: Target user ID
            event: Event name
            data: Event data
        """
        try:
            session_id = self.connected_users.get(user_id)
            
            if session_id:
                await self.sio.emit(event, data, room=session_id)
                logger.info(f"Emitted {event} to user {user_id}")
            else:
                logger.debug(f"User {user_id} not connected, skipping event {event}")
        
        except Exception as e:
            logger.error(f"Error emitting to user {user_id}: {e}")
    
    async def emit_to_all(self, event: str, data: Dict[str, Any]):
        """
        Emit an event to all connected users.
        
        Args:
            event: Event name
            data: Event data
        """
        try:
            await self.sio.emit(event, data)
            logger.info(f"Emitted {event} to all users")
        except Exception as e:
            logger.error(f"Error emitting to all: {e}")
    
    async def notify_new_notification(self, user_id: str, notification: Dict[str, Any]):
        """
        Send real-time notification to user.
        
        Args:
            user_id: Target user ID
            notification: Notification data
        """
        await self.emit_to_user(user_id, 'new_notification', notification)
    
    async def notify_validation_received(
        self,
        user_id: str,
        validator_name: str,
        action: str,
        claim_id: str
    ):
        """
        Notify claim owner of new validation in real-time.
        """
        await self.emit_to_user(user_id, 'validation_received', {
            'validator_name': validator_name,
            'action': action,
            'claim_id': claim_id,
            'message': f'{validator_name} {action} your claim'
        })
    
    async def notify_consensus_reached(
        self,
        user_id: str,
        claim_id: str,
        consensus_action: str,
        percentage: float
    ):
        """
        Notify claim owner of consensus result in real-time.
        """
        await self.emit_to_user(user_id, 'consensus_reached', {
            'claim_id': claim_id,
            'action': consensus_action,
            'percentage': percentage,
            'message': f'Consensus reached: {consensus_action} ({percentage}%)'
        })
    
    async def update_validation_count(self, claim_id: str, count: int):
        """
        Broadcast updated validation count for a claim.
        """
        await self.emit_to_all('validation_count_update', {
            'claim_id': claim_id,
            'count': count
        })
    
    async def update_consensus_percentage(
        self,
        claim_id: str,
        vouch_percentage: float,
        dispute_percentage: float,
        unsure_percentage: float
    ):
        """
        Broadcast updated consensus percentages for a claim.
        """
        await self.emit_to_all('consensus_update', {
            'claim_id': claim_id,
            'vouch': vouch_percentage,
            'dispute': dispute_percentage,
            'unsure': unsure_percentage
        })
    
    async def notify_badge_earned(self, user_id: str, badge_name: str, badge_id: str):
        """
        Notify user of badge achievement in real-time.
        """
        await self.emit_to_user(user_id, 'badge_earned', {
            'badge_name': badge_name,
            'badge_id': badge_id,
            'message': f'🏆 You earned the {badge_name} badge!'
        })
    
    async def notify_trust_score_updated(
        self,
        user_id: str,
        old_score: float,
        new_score: float,
        change: float
    ):
        """
        Notify user of trust score change in real-time.
        """
        await self.emit_to_user(user_id, 'trust_score_update', {
            'old_score': old_score,
            'new_score': new_score,
            'change': change,
            'message': f'Trust score {"increased" if change > 0 else "decreased"} by {abs(change):.1f}'
        })
    
    def register_user(self, user_id: str, session_id: str):
        """
        Register a user's WebSocket connection.
        
        Args:
            user_id: User ID
            session_id: Socket.IO session ID
        """
        self.connected_users[user_id] = session_id
        logger.info(f"User {user_id} registered with session {session_id}")
    
    def unregister_user(self, user_id: str):
        """
        Unregister a user's WebSocket connection.
        
        Args:
            user_id: User ID
        """
        if user_id in self.connected_users:
            del self.connected_users[user_id]
            logger.info(f"User {user_id} unregistered")
    
    def get_connected_count(self) -> int:
        """Get count of connected users."""
        return len(self.connected_users)


# Create singleton instance
websocket_service = WebSocketService()


# Socket.IO event handlers
@sio.event
async def connect(sid, environ):
    """Handle client connection."""
    logger.info(f"Client connected: {sid}")
    await sio.emit('connection_established', {'session_id': sid}, room=sid)


@sio.event
async def disconnect(sid):
    """Handle client disconnection."""
    logger.info(f"Client disconnected: {sid}")
    
    # Find and unregister user
    user_id = None
    for uid, session_id in websocket_service.connected_users.items():
        if session_id == sid:
            user_id = uid
            break
    
    if user_id:
        websocket_service.unregister_user(user_id)


@sio.event
async def authenticate(sid, data):
    """
    Authenticate user and register their connection.
    
    Client should send: { 'user_id': 'xxx', 'token': 'jwt_token' }
    """
    try:
        user_id = data.get('user_id')
        token = data.get('token')
        
        if not user_id:
            await sio.emit('auth_error', {'message': 'user_id required'}, room=sid)
            return
        
        # TODO: Verify JWT token
        # For now, we'll just register the user
        websocket_service.register_user(user_id, sid)
        
        await sio.emit('authenticated', {
            'user_id': user_id,
            'message': 'Successfully authenticated'
        }, room=sid)
        
        logger.info(f"User {user_id} authenticated with session {sid}")
    
    except Exception as e:
        logger.error(f"Authentication error: {e}")
        await sio.emit('auth_error', {'message': str(e)}, room=sid)


@sio.event
async def ping(sid):
    """Handle ping requests for keep-alive."""
    await sio.emit('pong', {}, room=sid)


@sio.event
async def get_status(sid):
    """Get WebSocket service status."""
    await sio.emit('status', {
        'connected_users': websocket_service.get_connected_count(),
        'session_id': sid
    }, room=sid)

File: frontend/src/services/websocketService.js
import { io } from 'socket.io-client';

/**
 * WebSocket Service
 * 
 * Real-time communication service using Socket.IO.
 * Features:
 * - Connect/disconnect management
 * - Authentication
 * - Real-time notifications
 * - Live validation updates
 * - Live consensus updates
 */
class WebSocketService {
  constructor() {
    this.socket = null;
    this.connected = false;
    this.authenticated = false;
    this.userId = null;
    this.listeners = {};
    this.reconnectAttempts = 0;
    this.maxReconnectAttempts = 5;
  }

  /**
   * Connect to WebSocket server
   * @param {string} url - WebSocket server URL (default: http://localhost:8000/ws)
   */
  connect(url = 'http://localhost:8000/ws') {
    if (this.socket && this.connected) {
      console.log('Already connected to WebSocket');
      return;
    }

    try {
      this.socket = io(url, {
        transports: ['websocket', 'polling'],
        reconnection: true,
        reconnectionDelay: 1000,
        reconnectionDelayMax: 5000,
        reconnectionAttempts: this.maxReconnectAttempts
      });

      this.setupEventHandlers();
      console.log('Connecting to WebSocket server...');
    } catch (error) {
      console.error('Failed to connect to WebSocket:', error);
    }
  }

  /**
   * Setup Socket.IO event handlers
   */
  setupEventHandlers() {
    // Connection events
    this.socket.on('connect', () => {
      console.log('Connected to WebSocket server');
      this.connected = true;
      this.reconnectAttempts = 0;
      
      // Trigger connect callback
      this.trigger('connect', { connected: true });
      
      // Re-authenticate if we have user ID
      if (this.userId) {
        this.authenticate(this.userId);
      }
    });

    this.socket.on('disconnect', (reason) => {
      console.log('Disconnected from WebSocket:', reason);
      this.connected = false;
      this.authenticated = false;
      this.trigger('disconnect', { reason });
    });

    this.socket.on('connect_error', (error) => {
      console.error('WebSocket connection error:', error);
      this.reconnectAttempts++;
      this.trigger('connect_error', { error, attempts: this.reconnectAttempts });
    });

    // Authentication events
    this.socket.on('connection_established', (data) => {
      console.log('Connection established:', data);
    });

    this.socket.on('authenticated', (data) => {
      console.log('Authenticated:', data);
      this.authenticated = true;
      this.trigger('authenticated', data);
    });

    this.socket.on('auth_error', (data) => {
      console.error('Authentication error:', data);
      this.authenticated = false;
      this.trigger('auth_error', data);
    });

    // Notification events
    this.socket.on('new_notification', (notification) => {
      console.log('New notification:', notification);
      this.trigger('new_notification', notification);
    });

    this.socket.on('validation_received', (data) => {
      console.log('Validation received:', data);
      this.trigger('validation_received', data);
    });

    this.socket.on('consensus_reached', (data) => {
      console.log('Consensus reached:', data);
      this.trigger('consensus_reached', data);
    });

    this.socket.on('badge_earned', (data) => {
      console.log('Badge earned:', data);
      this.trigger('badge_earned', data);
    });

    this.socket.on('trust_score_update', (data) => {
      console.log('Trust score update:', data);
      this.trigger('trust_score_update', data);
    });

    // Live update events
    this.socket.on('validation_count_update', (data) => {
      console.log('Validation count update:', data);
      this.trigger('validation_count_update', data);
    });

    this.socket.on('consensus_update', (data) => {
      console.log('Consensus update:', data);
      this.trigger('consensus_update', data);
    });

    // Status events
    this.socket.on('pong', () => {
      // Keep-alive response
    });

    this.socket.on('status', (data) => {
      console.log('WebSocket status:', data);
      this.trigger('status', data);
    });
  }

  /**
   * Authenticate with the server
   * @param {string} userId - User ID
   * @param {string} token - JWT token (optional)
   */
  authenticate(userId, token = null) {
    if (!this.socket || !this.connected) {
      console.error('Cannot authenticate: not connected');
      return;
    }

    this.userId = userId;
    
    // Get token from localStorage if not provided
    if (!token) {
      token = localStorage.getItem('token');
    }

    this.socket.emit('authenticate', {
      user_id: userId,
      token: token
    });
  }

  /**
   * Disconnect from WebSocket server
   */
  disconnect() {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
      this.connected = false;
      this.authenticated = false;
      this.userId = null;
      console.log('Disconnected from WebSocket');
    }
  }

  /**
   * Register an event listener
   * @param {string} event - Event name
   * @param {Function} callback - Callback function
   */
  on(event, callback) {
    if (!this.listeners[event]) {
      this.listeners[event] = [];
    }
    this.listeners[event].push(callback);
  }

  /**
   * Remove an event listener
   * @param {string} event - Event name
   * @param {Function} callback - Callback function
   */
  off(event, callback) {
    if (!this.listeners[event]) return;
    
    this.listeners[event] = this.listeners[event].filter(cb => cb !== callback);
  }

  /**
   * Trigger event listeners
   * @param {string} event - Event name
   * @param {any} data - Event data
   */
  trigger(event, data) {
    if (!this.listeners[event]) return;
    
    this.listeners[event].forEach(callback => {
      try {
        callback(data);
      } catch (error) {
        console.error(`Error in ${event} listener:`, error);
      }
    });
  }

  /**
   * Send ping to server (keep-alive)
   */
  ping() {
    if (this.socket && this.connected) {
      this.socket.emit('ping');
    }
  }

  /**
   * Get server status
   */
  getStatus() {
    if (this.socket && this.connected) {
      this.socket.emit('get_status');
    }
  }

  /**
   * Check if connected
   * @returns {boolean} Connection status
   */
  isConnected() {
    return this.connected;
  }

  /**
   * Check if authenticated
   * @returns {boolean} Authentication status
   */
  isAuthenticated() {
    return this.authenticated;
  }

  /**
   * Get connection info
   * @returns {object} Connection information
   */
  getConnectionInfo() {
    return {
      connected: this.connected,
      authenticated: this.authenticated,
      userId: this.userId,
      reconnectAttempts: this.reconnectAttempts
    };
  }

  /**
   * Start keep-alive ping interval
   * @param {number} interval - Ping interval in milliseconds (default: 30000)
   */
  startKeepAlive(interval = 30000) {
    if (this.keepAliveInterval) {
      clearInterval(this.keepAliveInterval);
    }

    this.keepAliveInterval = setInterval(() => {
      if (this.connected) {
        this.ping();
      }
    }, interval);
  }

  /**
   * Stop keep-alive ping interval
   */
  stopKeepAlive() {
    if (this.keepAliveInterval) {
      clearInterval(this.keepAliveInterval);
      this.keepAliveInterval = null;
    }
  }
}

// Export singleton instance
const websocketService = new WebSocketService();

// Auto-connect when token exists
const token = localStorage.getItem('token');
const user = localStorage.getItem('user');
if (token && user) {
  try {
    const userData = JSON.parse(user);
    websocketService.connect();
    // Will authenticate on connect event
  } catch (error) {
    console.error('Failed to parse user data:', error);
  }
}

export default websocketService;
